lib/skate/detours/route_segments.ex

```diff
@@ -0,0 +1,68 @@
+defmodule Skate.Detours.RouteSegments do
+  @moduledoc """
+  Break a route into segments based on `onnection_start` and `connection_end` points
+  """
+
+  defmodule Params do
+    @moduledoc false
+    @type t :: %__MODULE__{
+            connection_start: Util.Location.From.t(),
+            connection_end: Util.Location.From.t(),
+            shape: nonempty_list(Util.Location.From.t())
+          }
+    @enforce_keys [:connection_start, :connection_end, :shape]
+    defstruct [:connection_start, :connection_end, :shape]
+  end
+
+  defmodule Result do
+    @moduledoc false
+    @type t :: %__MODULE__{
+            before_detour: [Util.Location.From.t()],
+            detour: [Util.Location.From.t()],
+            after_detour: [Util.Location.From.t()]
+          }
+    @enforce_keys [:before_detour, :detour, :after_detour]
+    defstruct [:before_detour, :detour, :after_detour]
+  end
+
+  @doc """
+  Break a route shape into segments.
+
+  ## Examples
+      iex> alias Util.Location
+      iex> Skate.Detours.RouteSegments.route_segments(
+      ...>   %Skate.Detours.RouteSegments.Params{
+      ...>     shape: [Location.new(0, 0), Util.Location.new(0, 1), Util.Location.new(0, 2), Util.Location.new(0, 3)],
+      ...>     connection_start: Util.Location.new(0, 1),
+      ...>     connection_end: Util.Location.new(0, 2)
+      ...>   }
+      ...> )
+      %Skate.Detours.RouteSegments.Result{
+        before_detour: [%Util.Location{latitude: 0, longitude: 0}, %Util.Location{latitude: 0, longitude: 1}],
+        detour: [%Util.Location{latitude: 0, longitude: 1}, %Util.Location{latitude: 0, longitude: 2}],
+        after_detour: [%Util.Location{latitude: 0, longitude: 2}, %Util.Location{latitude: 0, longitude: 3}]
+      }
+  """
+  @spec route_segments(Params.t()) :: Result.t()
+  def route_segments(%__MODULE__.Params{
+        connection_start: connection_start,
+        connection_end: connection_end,
+        shape: shape
+      }) do
```

Comment:
**suggestion:** if you include a guard here I think it would provide a clearer and more up-front state when attempting to pass in an empty list, rather than having `Enum.min_by` to raise an exception.
```suggestion
      }) when length(shape) > 0 do
```

Reply:
Oh I accidentally lumped a few different things together into one commit, but https://github.com/mbta/skate/pull/2438/commits/434618f81afa36e2eaff1bafade928165ffd6e88 should address this